src/main/java/com/microsoft/sqlserver/jdbc/SQLServerXAResource.java

```diff
@@ -845,8 +845,19 @@ public void forget(Xid xid) throws XAException {
                 power = power * 256;
             }
             offset += 4;
-            int gid_len = (r.bData[offset++] & 0x00FF);
-            int bid_len = (r.bData[offset++] & 0x00FF);
+            int gid_len = (r.bData[offset] & 0x00FF);
+            if (offset++ > r.bData.length) {
+                MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_indexOutOfRange"));
+                Object[] msgArgs = {offset};
+                XAException xex = new XAException(form.format(msgArgs));
+                xex.errorCode = XAException.XAER_RMERR;
+                if (xaLogger.isLoggable(Level.FINER))
+                    xaLogger.finer(toString() + " exception:" + xex);
+                throw xex;
+            }
+            int bid_len = (r.bData[offset] & 0x00FF);
+            offset++;
```

Comment:
Same here, if we're really worried this may cause an out of index access then we can catch and rethrow it.

Reply:
I did this cos adding a try doesn't *fix* the error for Semmle. It's stupid. Ok I changed it